File: finding-seams/csharp/FindingSeams/RoomAvailability/RoomService.cs
namespace FindingSeams.RoomAvailability;

public record PlaceSessionResult(
    int SessionId,
    string RoomId,
    TimeSlot TimeSlot,
    string Status,
    string? Reason = null);

public class RoomService
{
    public PlaceSessionResult PlaceSessionInRoom(int sessionId, string roomId, TimeSlot timeSlot)
    {
        if (!RoomRegistries.Instance.IsAvailable(roomId, timeSlot))
        {
            return new PlaceSessionResult(sessionId, roomId, timeSlot, "REJECTED", "Room not available");
        }

        RoomRegistries.Instance.Reserve(roomId, timeSlot);
        return new PlaceSessionResult(sessionId, roomId, timeSlot, "ACCEPTED");
    }
}

File: finding-seams/kotlin/src/main/kotlin/findingseams/roomavailability/RoomService.kt
package findingseams.roomavailability

sealed class PlaceSessionResult {
    abstract val sessionId: Int
    abstract val roomId: String
    abstract val timeSlot: TimeSlot

    data class Accepted(
        override val sessionId: Int,
        override val roomId: String,
        override val timeSlot: TimeSlot,
        val status: String = "ACCEPTED",
    ) : PlaceSessionResult()

    data class Rejected(
        override val sessionId: Int,
        override val roomId: String,
        override val timeSlot: TimeSlot,
        val status: String = "REJECTED",
        val reason: String,
    ) : PlaceSessionResult()
}

class RoomService {
    fun placeSessionInRoom(sessionId: Int, roomId: String, timeSlot: TimeSlot): PlaceSessionResult {
        if (!roomRegistry.isAvailable(roomId, timeSlot)) {
            return PlaceSessionResult.Rejected(
                sessionId = sessionId,
                roomId = roomId,
                timeSlot = timeSlot,
                reason = "Room not available",
            )
        }
        roomRegistry.reserve(roomId, timeSlot)
        return PlaceSessionResult.Accepted(
            sessionId = sessionId,
            roomId = roomId,
            timeSlot = timeSlot,
        )
    }
}
